### PR TITLE
Vet clinic database: query and update animals table

### PR DESCRIPTION
**Pull request by Nahid1911**
### Requirements:

- Add a column species of type string to your animals table. Modify your schema.sql file.
- Insert provided data of animals into animals table and update  data.sql file.
- Pass a transaction to update the animals table by setting the values of species column.
- image
- Delete all records in the animals table, then roll back the transaction.
-image.
- Create a transaction with address several issues;
- image image 
imageimage

Answer to the questions: 

1. How many animals are there? 
image

2. How many animals have never tried to escape? 
image

3. What is the average weight of animals?
image

4. Who escapes the most, neutered or not neutered animals?
image

5. What is the minimum and maximum weight of each type of animal?
image

6. What is the average number of escape attempts per animal type of those born between 1990 and 2000?
image

**For details requirements please click [HERE](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md)**

### CODE REVIEW

File: data.sql [hunk 1]
```diff
@@ -5,3 +5,12 @@ VALUES ('Agumon', '2020-02-03', 0, true, 10.23),
 ('Gabumon', '2018-11-15', 2, true, 8.00),
 ('Pickachu', '2021-01-07', 1, false, 15.04),
 ('Devimon', '2017-05-12', 5, true, 11.00);
+
+INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
+VALUES ('Charmander', '2020-02-08', 2, true, 5.70),
```
Comment (topeogunleye) [inline data.sql]: Hello, thank you for submitting your code for review. I appreciate your hard work and dedication. 😊

I have checked your code and I have some feedback for you. Please read it carefully and try to improve your code accordingly.

First of all, your code does not meet the project requirements. The project requires you to insert the following data into the animals table:

Animal: His name is Charmander. He was born on Feb 8th, 2020, and currently weighs -11kg. He is not neutered and he has never tried to escape.

However, you inserted different data for the escape_attempts, neutered, and weight_kg columns:

```sql
INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
VALUES ('Charmander', '2020-02-08', 2, true, 5.70);
```

This is a serious mistake, because it means that you did not follow the instructions and that your data is incorrect. You should always make sure that you insert the correct data into the table, otherwise you will get wrong results and errors.

I hope this feedback was helpful and constructive. If you have any questions or doubts, please feel free to ask me. I'm here to help you learn and improve your skills. 😊

Please revise your code and resubmit it for review. I'm looking forward to seeing your progress. 👍

Have a great day! 😊

File: data.sql [hunk 2]
```diff
@@ -5,3 +5,12 @@ VALUES ('Agumon', '2020-02-03', 0, true, 10.23),
...
+('Palantmon', '2021-11-15', 2, true, 5.70),
```
Comment (topeogunleye) [inline data.sql]: Hello, thank you for submitting your code for review. I appreciate your hard work and dedication. 😊

I have checked your code and I have some feedback for you. Please read it carefully and try to improve your code accordingly.

First of all, your code does not meet the project requirements. The project requires you to insert the following data into the animals table:

Animal: Her name is Plantmon. She was born on Nov 15th, 2021, and currently weighs -5.7kg. She is neutered and she has tried to escape 2 times.

However, you inserted different data for the name, date_of_birth, and weight_kg columns:

```sql
('Palantmon', '2021-11-15', 2, true, 5.70)
```

This is a serious mistake, because it means that you did not follow the instructions and that your data is incorrect. You should always make sure that you insert the correct data into the table, otherwise you will get wrong results and errors.

This will insert one row of data into the animals table, with the specified values for each column:

```sql
INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
VALUES ('Plantmon', '2021-11-15', 2, true, -5.70);
```


I hope this feedback was helpful and constructive. If you have any questions or doubts, please feel free to ask me. I'm here to help you learn and improve your skills. 😊

Please revise your code and resubmit it for review. I'm looking forward to seeing your progress. 👍

Have a great day! 😊

File: data.sql [hunk 3]
```diff
@@ -5,3 +5,12 @@ VALUES ('Agumon', '2020-02-03', 0, true, 10.23),
...
+('Squirtle', '1993-04-02', 3, false, 12.13),
```
Comment (topeogunleye) [inline data.sql]: Hello, thank you for submitting your code for review. I appreciate your hard work and dedication. 😊

I have checked your code and I have some feedback for you. Please read it carefully and try to improve your code accordingly.

First of all, your code meets most of the project requirements. The project requires you to insert the following data into the animals table:

Animal: His name is Squirtle. He was born on Apr 2nd, 1993, and currently weighs -12.13kg. He was not neutered and he has tried to escape 3 times.

You inserted the correct data for the name, date_of_birth, escape_attempts, and neutered columns:

```sql
INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
VALUES ('Squirtle', '1993-04-02', 3, false, 12.13);
```

This is good, because it means that you followed the instructions and that your data is correct. You should always make sure that you insert the correct data into the table, otherwise you will get wrong results and errors.

However, you inserted a different data for the weight_kg column:

```sql
VALUES ('Squirtle', '1993-04-02', 3, false, 12.13);
```

This is a minor mistake, but it can affect the accuracy of your data. 

To fix this mistake, you should change the value of the weight_kg column to -12.13:

```sql
VALUES ('Squirtle', '1993-04-02', 3, false, -12.13);
```

This will insert the correct value for the weight_kg column into the animals table.

In summary, your code has only one error and meets most of the project requirements. You need to fix this issue and make sure that your code is correct and consistent. I suggest that you review the SQL syntax and data types and practice more with inserting data into tables.

I hope this feedback was helpful and constructive. If you have any questions or doubts, please feel free to ask me. I'm here to help you learn and improve your skills. 😊

Please revise your code and resubmit it for review. I'm looking forward to seeing your progress. 👍

Have a great day! 😊

File: data.sql [hunk 4]
```diff
@@ -5,3 +5,12 @@ VALUES ('Agumon', '2020-02-03', 0, true, 10.23),
...
+('Angemon', '2005-06-12', 1, true, 45.00),
```
Comment (topeogunleye) [inline data.sql]: - [ ] As I mentioned in the previous statement, please you inserted a different data for the weight_kg column. it should be -45.00 instead

File: data.sql [hunk 5]
```diff
@@ -5,3 +5,12 @@ VALUES ('Agumon', '2020-02-03', 0, true, 10.23),
...
+('Boarmon', '2005-06-07', 7, true, 20.40),
+('Blossom', '1998-10-13', 3, true, 17.00),
+('Ditto', '2022-05-14', 4, true, 17.00);
```
Comment (topeogunleye) [inline data.sql]: - [ ] As I mentioned in the previous statement, please you have inserted a different data for the weight_kg column. it should be 22.00 instead

File: queries.sql [hunk 6]
```diff
@@ -15,3 +15,44 @@ SELECT * FROM animals WHERE neutered = true;
 SELECT * FROM animals WHERE name NOT LIKE 'Gabumon';
 
 SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
+
+BEGIN;
+ SAVEPOINT sp1;
+ UPDATE animals
+ SET species = 'unspecified';
+ ROLLBACK TO sp1;
+COMMIT;
```
Comment (topeogunleye) [inline queries.sql]: Hello, thank you for submitting your code for review. I appreciate your hard work and dedication. 😊

I have checked your code and I have some feedback for you. Please read it carefully and try to improve your code accordingly.

First of all, your code meets some of the project requirements. The project requires you to update the animals table by setting the species column to unspecified, then roll back the change, inside a transaction. You used the BEGIN, UPDATE, ROLLBACK, and COMMIT statements correctly to perform this task:

```sql
BEGIN;
 SAVEPOINT sp1;
 UPDATE animals
 SET species = 'unspecified';
 ROLLBACK TO sp1;
COMMIT;
```

This is good, because it means that you understand how to use transactions and rollbacks to control the changes in the database. You should always use transactions and rollbacks when you want to test or undo your changes without affecting the data permanently.

However, you missed one important step in the project requirements. You did not verify that the change was made and rolled back by using the SELECT statement:

```sql
SELECT species from animals;
```

This is a crucial step, because it allows you to check the results of your update and rollback statements. You should always use the SELECT statement to query the data before and after your changes, to make sure that they work as expected.

To fix this mistake, you should add the SELECT statement after the UPDATE and ROLLBACK statements, like this:

```sql
BEGIN;
 SAVEPOINT sp1;
 UPDATE animals
 SET species = 'unspecified';
 SELECT species from animals; -- verify that change was made
 ROLLBACK TO sp1;
 SELECT species from animals; -- verify that change was undone
COMMIT;
```

This will show you the values of the species column before and after your update and rollback statements. You should see that the values are changed to 'unspecified' after the update, and restored to their original values after the rollback.

In summary, your code has only one error and meets some of the project requirements. You need to fix this issue and make sure that your code is correct and complete. I suggest that you review the SQL syntax and commands and practice more with transactions and rollbacks.

I hope this feedback was helpful and constructive. If you have any questions or doubts, please feel free to ask me. I'm here to help you learn and improve your skills. 😊

Please revise your code and resubmit it for review. I'm looking forward to seeing your progress. 👍

Have a great day! 😊

File: queries.sql [hunk 7]
```diff
@@ -15,3 +15,44 @@ SELECT * FROM animals WHERE neutered = true;
...
+
+BEGIN;
+UPDATE animals
+SET species = CASE WHEN name LIKE '%mon' THEN 'digimon' ELSE 'pokemon' END;
+COMMIT;
```
Comment (topeogunleye) [inline queries.sql]: Hello, thank you for submitting your code for review. I appreciate your hard work and dedication. 😊

I have checked your code and I have some feedback for you. Please read it carefully and try to improve your code accordingly.

First of all, your code meets some of the project requirements. The project requires you to update the animals table by setting the species column to digimon for all animals that have a name ending in mon, and to pokemon for all animals that don't have species already set, inside a transaction. You used the BEGIN, UPDATE, and COMMIT statements correctly to perform this task:

```sql
BEGIN;
UPDATE animals
SET species = CASE WHEN name LIKE '%mon' THEN 'digimon' ELSE 'pokemon' END;
COMMIT;
```

This is good, because it means that you understand how to use transactions and updates to modify the data in the database. You should always use transactions and updates when you want to change the data in a table, according to some conditions.

However, you missed some important steps in the project requirements. You did not specify the condition for updating the species column to pokemon only for the animals that don't have species already set. You also did not verify that the changes were made and persisted after the commit. These are crucial steps, because they allow you to check the results and the effects of your update and commit statements. You should always use the SELECT statement to query the data before and after your changes, to make sure that they work as expected.

To fix these mistakes, you should add the following steps to your code:

- Add another condition to the CASE expression, using the WHEN clause, to update the species column to pokemon only when it is NULL:

```sql
SET species = CASE WHEN name LIKE '%mon' THEN 'digimon' WHEN species IS NULL THEN 'pokemon' END;
```

This will update the species column to digimon for all animals that have a name ending in mon, and to pokemon for all animals that don't have species already set, leaving the other values unchanged.

- Add a SELECT statement before the UPDATE statement, to show the original values of the species column:

```sql
SELECT name, species from animals;
```

This will show you the name and species of each animal before the update.

- Add a SELECT statement after the UPDATE statement, to show the updated values of the species column:

```sql
SELECT name, species from animals;
```

This will show you the name and species of each animal after the update.

- Add a SELECT statement after the COMMIT statement, to show that the changes persist after the commit:

```sql
SELECT name, species from animals;
```

This will show you the name and species of each animal after the commit.

Your final code should look like this:

```sql
BEGIN; -- start transaction
SELECT name, species from animals; -- show original values
UPDATE animals
SET species = CASE WHEN name LIKE '%mon' THEN 'digimon' WHEN species IS NULL THEN 'pokemon' END; -- make changes
SELECT name, species from animals; -- show updated values
COMMIT; -- end transaction
SELECT name, species from animals; -- show persisted values
```

This will update and commit the changes in the animals table as required by the project.

In summary, your code has some errors and misses some steps in meeting the project requirements. You need to fix these issues and make sure that your code is correct and complete. I suggest that you review the SQL syntax and commands and practice more with transactions and updates.

I hope this feedback was helpful and constructive. If you have any questions or doubts, please feel free to ask me. I'm here to help you learn and improve your skills. 😊

Please revise your code and resubmit it for review. I'm looking forward to seeing your progress. 👍

Have a great day! 😊

File: queries.sql [hunk 8]
```diff
@@ -15,3 +15,44 @@ SELECT * FROM animals WHERE neutered = true;
...
+SAVEPOINT sp1;
+UPDATE animals SET weight_kg = weight_kg * -1;
+ROLLBACK TO sp1;
+UPDATE animals SET weight_kg = CASE WHEN weight_kg < 0 THEN weight_kg * -1 ELSE weight_kg END;
+COMMIT;
+
+
+SELECT COUNT(name) AS number_of_animals FROM animals;
+
+SELECT COUNT(name) AS tried_to_escape FROM animals WHERE escape_attempts > 0;
```
Comment (topeogunleye) [inline queries.sql]: Hello, thank you for submitting your code for review. I appreciate your hard work and dedication. 😊

I have checked your code and I have some feedback for you. Please read it carefully and try to improve your code accordingly.

First of all, your code does not meet the project requirements. The project requires you to write queries to answer the following question:

How many animals have never tried to escape?

However, you wrote a query to answer a different question:

How many animals have tried to escape at least once?

```sql
SELECT COUNT(name) AS tried_to_escape FROM animals WHERE escape_attempts > 0;
```

This is a mistake, because it means that you did not follow the instructions and that your query is incorrect. You should always make sure that you write queries that match the question, otherwise you will get wrong results and errors.

To fix this mistake, you should write a query that counts the number of animals that have zero escape attempts, like this:

```sql
SELECT COUNT(name) AS never_tried_to_escape FROM animals WHERE escape_attempts = 0;
```

This will count the number of animals that have never tried to escape and give it an alias of never_tried_to_escape.

In summary, your code has one error and does not meet the project requirements. You need to fix this issue and make sure that your code is correct and consistent. I suggest that you review the SQL syntax and operators and practice more with writing queries.

I hope this feedback was helpful and constructive. If you have any questions or doubts, please feel free to ask me. I'm here to help you learn and improve your skills. 😊

Please revise your code and resubmit it for review. I'm looking forward to seeing your progress. 👍

Have a great day! 😊

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @Nahid1911, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 

- You added a new column called species with a string type to the animals table in the schema.sql file. ✔️ 
- You used the ALTER TABLE statement to add a new column to an existing table. ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
👉 One optional suggestion for improvement is to follow the convention for naming your branch. Usually, branch names should use the past tense verb to indicate that the feature has been implemented. For example, instead of `updateAnimTbl`, you could use `updatedAnimTbl`. This would make your branch name more consistent and clear for future pull requests.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: data.sql (hunk 1 above)
Comment (Nahid1911) [reply data.sql]: @topeogunleye Thank you for the correction
Data has been updated accordingly.

File: data.sql (hunk 2 above)
Comment (Nahid1911) [reply data.sql]: @topeogunleye Thank you for the correction, I must admit that I didn't initially consider the possibility of negative weights, and your clarification helped me address this oversight promptly.

Data Updated accordingly.

File: data.sql (hunk 3 above)
Comment (Nahid1911) [reply data.sql]: @topeogunleye Thank you for the correction, I must admit that I didn't initially consider the possibility of negative weights, and your clarification helped me address this oversight promptly.

Data Updated accordingly.

File: data.sql (hunk 4 above)
Comment (Nahid1911) [reply data.sql]: @topeogunleye Thank you for the correction, I must admit that I didn't initially consider the possibility of negative weights, and your clarification helped me address this oversight promptly.

Data Updated accordingly.

File: data.sql (hunk 5 above)
Comment (Nahid1911) [reply data.sql]: @topeogunleye for addressing this in your comments.

Data updated accordingly.

File: queries.sql (hunk 6 above)
Comment (Nahid1911) [reply queries.sql]: @topeogunleye Thank you for addressing the issue, 

The code updated accordingly.

File: queries.sql (hunk 7 above)
Comment (Nahid1911) [reply queries.sql]: @topeogunleye Thank you for the correction,

Code Updated accordingly.

File: queries.sql (hunk 8 above)
Comment (Nahid1911) [reply queries.sql]: @topeogunleye Thank you for the correction, I must admit that i have learnt a lot from your review, But repeating some lines in every comments and my correct code in your comments made a little confusion and it is complected to find the exact problems and suggested changes. .

Data Updated accordingly.

Comment (Nahid1911) [reply queries.sql]: @topeogunleye Thank you for the correction.
Code updated accordingly.

Comment (Nahid1911) [reply queries.sql]: @topeogunleye 
I want to express my gratitude for taking the time to review my code. Your corrections and suggestions have been immensely valuable, and I have learned a lot from your review.

However, I would like to offer a suggestion to improve the clarity of the feedback. I noticed that some lines in your comments, including my corrected code, are repeated across different sections. As a result, it made it a little confusing for me to identify the exact problems and the corresponding suggested changes.

In the future, if it's possible, could you consider consolidating similar comments or using references to specific sections of the code to provide more focused feedback? This would greatly help in understanding and addressing the issues efficiently.

Once again, I truly appreciate your effort and expertise in reviewing my code. Your feedback has been instrumental in my growth as a developer.

Thanking You
Nahid Raihan

**Review by shella12 — APPROVED** [review]
Hi @Nahid1911, 


## Status : APPROVED 🥳 🥳 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- data file is corrected ✔️ 
- Schema file s corrected ✔️ 
- Ran all queries  ✔️ 
- Descriptive PR ✔️ 
- Implemented all required changes 👏 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
